File: client/java/src/main/java/com/be/mesh/client/struct/Digest.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.struct;

import com.be.mesh.client.mpc.Mesh;
import com.be.mesh.client.tool.Tool;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;

import java.io.Serializable;
import java.util.Optional;

/**
 * @author coyzeng@gmail.com
 */
@Getter
@Slf4j(topic = "digest")
public class Digest implements Serializable {

    private static final long serialVersionUID = -135075634815713279L;
    private final String traceId = Mesh.context().getTraceId();
    private final String spanId = Mesh.context().getSpanId();
    private final String mode = String.valueOf(Mesh.context().getRunMode().getMode());
    private final String cni = Optional.ofNullable(Mesh.context().getConsumer().getNodeId()).orElse("");
    private final String cii = Optional.ofNullable(Mesh.context().getConsumer().getInstId()).orElse("");
    private final String cip = Optional.ofNullable(Mesh.context().getConsumer().getIp()).orElse("");
    private final String chost = Optional.ofNullable(Mesh.context().getConsumer().getHost()).orElse("");
    private final String pni = Optional.ofNullable(Mesh.context().getProvider().getNodeId()).orElse("");
    private final String pii = Optional.ofNullable(Mesh.context().getProvider().getInstId()).orElse("");
    private final String pip = Optional.ofNullable(Mesh.context().getProvider().getIp()).orElse("");
    private final String phost = Optional.ofNullable(Mesh.context().getProvider().getHost()).orElse("");
    private final String urn = Mesh.context().getUrn();
    private final long now = System.currentTimeMillis();

    public void print(String pattern, String code) {
        log.info("{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{}",
                traceId,
                spanId,
                Mesh.context().getTimestamp(),
                now,
                System.currentTimeMillis() - Mesh.context().getTimestamp(),
                System.currentTimeMillis() - now,
                mode,
                pattern,
                cni,
                cii,
                pni,
                pii,
                cip,
                pip,
                chost,
                phost,
                Tool.anyone(Mesh.context().getAttribute(Mesh.REMOTE), Tool.MESH_ADDRESS.get().any()),
                urn,
                code);
    }

    public void mdc() {
        // Refresh slf4j MDC
        MDC.put("tid", traceId);
        MDC.put("sid", spanId);
        MDC.put("mod", mode);
        MDC.put("cni", cni);
        MDC.put("cii", cii);
        MDC.put("pni", pni);
        MDC.put("pii", pii);
    }
}

File: client/python/mesh/mpc/digest.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
import time

import mesh.log as log
import mesh.tool as tool
from mesh.context import Mesh


class Digest:

    def __init__(self):
        self.trace_id = Mesh.context().get_trace_id()
        self.span_id = Mesh.context().get_span_id()
        self.mode = Mesh.context().get_run_mode().value
        self.cni = tool.anyone(Mesh.context().get_consumer().node_id, '')
        self.cii = tool.anyone(Mesh.context().get_consumer().inst_id, '')
        self.cip = tool.anyone(Mesh.context().get_consumer().ip, '')
        self.chost = tool.anyone(Mesh.context().get_consumer().host, '')
        self.pni = tool.anyone(Mesh.context().get_provider().node_id, '')
        self.pii = tool.anyone(Mesh.context().get_provider().inst_id, '')
        self.pip = tool.anyone(Mesh.context().get_provider().ip, '')
        self.phost = tool.anyone(Mesh.context().get_provider().host, '')
        self.urn = Mesh.context().get_urn()
        self.now = int(time.time() * 1000)

    def write(self, pattern: str, code: str):
        now = int(time.time() * 1000)
        log.info(
            f"{self.trace_id},"
            f"{self.span_id},"
            f"{Mesh.context().get_timestamp()},"
            f"{self.now},"
            f"{now - Mesh.context().get_timestamp()},"
            f"{now - self.now},"
            f"{self.mode},"
            f"{pattern},"
            f"{self.cni},"
            f"{self.cii},"
            f"{self.pni},"
            f"{self.pii},"
            f"{self.cip},"
            f"{self.pip},"
            f"{self.chost},"
            f"{self.phost},"
            f"{tool.anyone(Mesh.context().get_attribute(Mesh.REMOTE), tool.get_mesh_address().any())},"
            f"{self.urn},{code}")
